qa/tasks/rsync.py

```diff
@@ -0,0 +1,276 @@
+            self.my_mnt.run_shell(["mkdir", "{}".format(self.source_dir)])
+            should_stop = False
+
+        #Check for source directory exists
+        while not self.check_if_dir_exists(self.source_dir):
+            time.sleep(5) # if source dorectory not exists wait for 5s and poll
+            iteration += 1
+            if iteration > 5:
+                assert self.check_if_dir_exists(self.source_dir), 'assert, source Directory does not exists'
+
+        # Start observing the event started by workunit task.
+        if self.work_unit:
+            should_stop = self.ctx.workunit_state.start_observing()
+
+        while not (should_stop or self.stopping.is_set()):
+
+            # rsync data from snapshot. snap is created using workunit IO data
+            if self.work_unit and self.snap_enable:
+
+                snap_shot = self.source_dir + '.snap/snap' + '{}'.format(iteration)
+
+                # Create Snapshot
+                self.my_mnt.run_shell(["mkdir", "{}".format(snap_shot)])
+                iteration += 1
+
+                self.my_mnt.run_shell(["rsync", "-azvh", "{}".format(snap_shot), "rsyncdir/dir1/"])
+
+                # Delete snapshot
+                self.my_mnt.run_shell(["rmdir", "{}".format(snap_shot)])
+
+                # Check for even handler stop message
+                finished = self.ctx.workunit_state.observer_should_stop()
+
+            # rsync data from snapshot, snap is created using written pattern data
+            elif self.snap_enable:
+                # Create file and add data to the file
+                self.my_mnt.write_test_pattern("{}/file_a".format(self.source_dir), self.file_size * 1024 * 1024)
+                snap_shot = self.data_dir + '/.snap/snap' + '{}'.format(iteration)
+
+                # Create Snapshot
+                self.my_mnt.run_shell(["mkdir", "{}".format(snap_shot)])
+                iteration += 1
+
+                self.my_mnt.run_shell(["rsync", "-azvh", "{}".format(snap_shot), "rsyncdir/dir{}/".format(iteration)])
+
+                # Delete snapshot
+                self.my_mnt.run_shell(["rmdir", "{}".format(snap_shot)])
+
+                # Delete the file created in data directory
+                self.my_mnt.run_shell(["rm", "-f", "{}/file_a".format(self.source_dir)])
+
+            # rsync data from workunit IO data
+            elif self.work_unit:
+                try:
+                    self.my_mnt.run_shell(["rsync", "-azvh", "{}".format(self.source_dir), "rsyncdir/dir1/"])
+                except CommandFailedError as e:
+                    if e.exitstatus == 24:
+                        log.info("Some files vanished before they could be transferred")
+                    else:
+                        raise
+                finally:
+                    # Check for event handler stop message
+                    finished = self.ctx.workunit_state.observer_should_stop()
```

Comment:
This should be `should_stop = ...` yes?

Reply:
Yes. That's correct.. I'm a bit stumped how the test passed. I think stopping(one of the parameters in the while loop along with should_stop) gets set due to timeout( i.e start_rsync_thread.stop() is called). Changed it.